File: leetcode/145.binary-tree-postorder-traversal/ts/postordertraversal_recur.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const postorderTraversal = (root: TreeNode | null): number[] => {
    const res: number[] = []
    traversal(root, res)
    return res
}

const traversal = (root: TreeNode | null, res: number[]) => {
    if (root === null) return

    traversal(root.left, res)
    traversal(root.right, res)
    res.push(root.val)
}

export default postorderTraversal

File: leetcode/145.binary-tree-postorder-traversal/go/postordertraversal_recur.go
package postordertraversal

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func postorderTraversal(root *TreeNode) []int {
    res := make([]int, 0)

	return traversal(root, res)
}

func traversal(root *TreeNode, res []int) []int {
	if root == nil {
		return res
	}

	res = traversal(root.Left, res)
	res = traversal(root.Right, res)

	res = append(res, root.Val)

	return res
}
